File: src/main/java/play/pluv/login/application/LoginService.java
package play.pluv.login.application;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import play.pluv.member.domain.Member;
import play.pluv.oauth.application.SocialLoginClientComposite;
import play.pluv.oauth.domain.OAuthMemberInfo;
import play.pluv.playlist.domain.MusicStreaming;

@Service
@RequiredArgsConstructor
public class LoginService {

  private final RegisterReader registerReader;
  private final RegisterUpdater registerUpdater;
  private final SocialLoginClientComposite socialLoginClientComposite;

  @Transactional
  public Long createToken(final MusicStreaming serverType, final String key) {
    final OAuthMemberInfo memberInfo = socialLoginClientComposite
        .fetchMemberInfo(serverType, key);

    final Member member = registerReader.findByOAuthMemberInfo(memberInfo)
        .orElseGet(() -> registerUpdater.registerNewMember(memberInfo));

    return member.getId();
  }

  @Transactional
  public void addOtherLoginWay(
      final MusicStreaming serverType, final Long memberId, final String key
  ) {
    final OAuthMemberInfo memberInfo = socialLoginClientComposite
        .fetchMemberInfo(serverType, key);

    registerUpdater.addOtherLoginSource(memberId, memberInfo);
  }
}

File: src/main/kotlin/play/pluv/login/application/LoginService.kt
package play.pluv.login.application

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import play.pluv.login.domain.TesterLoginValidator
import play.pluv.oauth.application.SocialLoginClientComposite
import play.pluv.playlist.domain.MusicStreaming

@Service
open class LoginService(
    private val registerUpdater: RegisterUpdater,
    private val registerReader: RegisterReader,
    private val socialLoginClientComposite: SocialLoginClientComposite,
    private val testerLoginValidator: TesterLoginValidator
) {

    @Transactional
    open fun registerAndGetMemberId(serverType: MusicStreaming, key: String): Long {
        val memberInfo = socialLoginClientComposite.fetchMemberInfo(serverType, key)
        val member = registerReader.findByOAuthMemberInfo(memberInfo)
            ?: registerUpdater.registerNewMember(oAuthMemberInfo = memberInfo)

        return member.identifier
    }

    @Transactional
    open fun addOtherLoginWay(serverType: MusicStreaming, memberId: Long, key: String) {
        val memberInfo = socialLoginClientComposite.fetchMemberInfo(serverType, key)
        registerUpdater.addOtherLoginSource(memberId, memberInfo)
    }

    @Transactional(readOnly = true)
    open fun getLoginTypes(memberId: Long): List<MusicStreaming> {
        return registerReader.findMemberSocialLoginIds(memberId)
            .map { it.source }
            .toList()
    }

    @Transactional(readOnly = true)
    open fun getTesterId(id: String, password: String): Long {
        val testerId = testerLoginValidator.getTesterId(id, password)
        return registerReader.findById(testerId).identifier
    }
}